docs(throttle): first draft of docs for the throttler module (#1370)
nestjs/docs.nestjs.com: content/security/rate-limiting.md

### Rate Limiting

A common technique to protect applications from brute-force attacks is **rate-limiting**. To get started, you'll need to install the `@nestjs/throttler` package.

```bash
$ npm i --save @nestjs/throttle
```

Once the installation is complete, the `ThrottlerModule` can be configured as any other Nest package with `forRoot` or `forRootAsync` methods.

```typescript
@Module({
  imports: [
    ThrottlerModule.forRoot({
      ttl: 60,
      limit: 10,
    }),
  ]
})
export class AppModule {}
```

The above will set the global options for the `ttl`, the time to live, and the `limit`, the maximum number of requests within the ttl, for the routes of your application that are guarded.

> Once the module has been imported, you can then choose how you would like to bind the `ThrottlerGuard`. Any kind of binding as mentioned in the [guards](https://docs.nestjs.com/guards) section is fine. If you wanted to bind the guard globally, for example, you could do so but adding this provider to any module

Reviewer: ```suggestion
Once the module has been imported, you can then choose how you would like to bind the `ThrottlerGuard`. Any kind of binding as mentioned in the [guards](https://docs.nestjs.com/guards) section is fine. If you wanted to bind the guard globally, for example, you could do so by adding this provider to any module
```

Reviewer 2: 👍 

+ let's add : or . by the end of the sentence

```typescript
{
  provide: APP_GUARD,
  useClass: ThrottlerGuard
}
```

#### Customization

> There may be a time where you want to bind the guard to a controller or globally, but want to avoid rate limiting one or more of your endpoints. For that, you can use the `@SkipThrottle()` decorator, to negate the throttler for an entire class or a single route. The `@SkipThrottle()` decorator can also take in a boolean for if there is a case where you want to exclude _most_ of a controller, but not every route.

Reviewer: ```suggestion
There may be a time where you want to bind the guard to a controller or globally, but want to disable rating limiting for one or more of your endpoints. For that, you can use the `@SkipThrottle()` decorator, to negate the throttler for an entire class or a single route. The `@SkipThrottle()` decorator can also take in a boolean for if there is a case where you want to exclude _most_ of a controller, but not every route.
```

#### GraphQL

Currently, only GraphQL with Express is supported, but Fastify support is coming as well. This module makes use of setting headers through the `res` object and reading headers through the `req` object of Express. To make sure these are available, when configuring your GraphQLModule, make sure the option `context: ({{ '{' }} req, res {{ '}' }}) => ({{ '{' }} req, res {{ '}' }})` is set.

#### Configuration

The following options are valid for the `ThrottlerModule`

<table>
  <tr>
    <td><code>ttl</code></td>
    <td>the number of seconds that each request will last in storage</td>
  </tr>
  <tr>
    <td><code>limit</code></td>
    <td>the maximum number of requests within the TTL limit</td>
  </tr>
  <tr>
    <td><code>ignoreUserAgents</code></td>
    <td>an array of regular expressions of user-agents to ignore when it comes to throttling requests</td>
  </tr>
  <tr>
    <td><code>storage</code></td>
    <td> the storage setting for how to keep track of the requests</td>
  </tr>
</table>

> #### Async Configuration

Reviewer: Wouldn't it be better to have the `async configuration` right below the regular `sync configuration`. And then all the configuration stuff? I think, this would enhance readability of this section dramatically.

Author: This is how most of Nest's docs are set up. If you take a look at the `Database` page you'll see a similar structure.

Reviewer: yeah, i know.. but i find this quite confusing :D in my opinion it would be better to have everything in one place.. but you're right - keeping the same structure on every page may also be a good idea :laughing:

Reviewer 2: agree with @jmcdo29, let's stay consistent :)

### Rate Limiting

> A common technique to protect applications from brute-force attacks is **rate-limiting**. To get started, you'll need to install the `@nestjs/throttler` package.

Reviewer 2: `@nestjs/throttler` here, `@nestjs/throttle` below

```typescript
@Module({
  imports: [
    ThrottlerModule.forRoot({
      ttl: 60,
      limit: 10,
    }),
  ]
})
export class AppModule {}
```

The above will set the global options for the `ttl`, the time to live, and the `limit`, the maximum number of requests within the ttl, for the routes of your application that are guarded.

Once the module has been imported, you can then choose how you would like to bind the `ThrottlerGuard`. Any kind of binding as mentioned in the [guards](https://docs.nestjs.com/guards) section is fine. If you wanted to bind the guard globally, for example, you could do so but adding this provider to any module

```typescript
{
  provide: APP_GUARD,
  useClass: ThrottlerGuard
}
```

#### Customization

There may be a time where you want to bind the guard to a controller or globally, but want to avoid rate limiting one or more of your endpoints. For that, you can use the `@SkipThrottle()` decorator, to negate the throttler for an entire class or a single route. The `@SkipThrottle()` decorator can also take in a boolean for if there is a case where you want to exclude _most_ of a controller, but not every route.

> There is also the `@Throttle()` decorator which can be used to override the `limit` and `ttl` set in the global module, to give tighter or looser security options. This decorator can be used on a class or a function as well. The order for this decorator does matter, as the arguments are in the order of `limit` `ttl`.

Reviewer 2: ```suggestion
There is also the `@Throttle()` decorator which can be used to override the `limit` and `ttl` set in the global module, to give tighter or looser security options. This decorator can be used on a class or a function as well. The order for this decorator does matter, as the arguments are in the order of `limit, ttl`.
```

Reviewer 3: I personally wouldn't change this part. I think `limit` `ttl` indicates a bit more clearly that these are 2 separate configurable properties.

#### Websockets

> This module _can_ work with websockets, but it requires some class extension. You can extend the `ThrottlerGuard` and override the `handleRequest` method like so:

Reviewer 2: ```suggestion
This module can work with websockets, but it requires some class extension. You can extend the `ThrottlerGuard` and override the `handleRequest` method like so:
```

```typescript
@Injectable()
export class WsThrottlerGuard extends ThrottlerGuard {
  async handleRequest(context: ExecutionContext, limit: number, ttl: number): Promise<boolean> {
    const client = context.switchToWs().getClient();
    const ip = client.conn.remoteAddress; 
    const key = this.generateKey(context, ip);
    const ttls = await this.storageService.getRecord(key);

    if (ttls.length >= limit) {
      throw new ThrottlerException();
    }

    await this.storageService.addRecord(key, ttl);
    return true;
  }
}
```

> info **Hint** If you are using the `@nestjs/platform-ws` package you can use  `client._socket.remoteAddress` instead.

#### GraphQL

The `ThrottlerGuard` can also be used to work with GraphQL requests. Again, the guard can be extended, but this tme the `getRequestResponse` method will be overridden

```typescript
@Injectable()
export class GqlThrottlerGuard extends ThrottlerGuard {

>   getRequestResponse(context: ExecutionContext): { req: Record<string, any>, res: Record<string, any> } {

Reviewer 2: ```suggestion
  getRequestResponse(context: ExecutionContext) {
```

    const gqlCtx = GqlExecutionContext.create(context);
    const ctx = gql.getContext();
    return { req, ctx.req, res: ctx.res }
  }
}
```

#### Configuration

> The following options are valid for the `ThrottlerModule`

Reviewer 2: ```suggestion
The following options are valid for the `ThrottlerModule`:
```

One approach would be to use a factory function:

```typescript
@Module({
  imports: [
    ThrottlerModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (config: ConfigService) => ({
        ttl: config.get('THROTTLE_TTL'),
        limit: config.get('THROTTLE_LIMIT'),
      }),
    }),
  ],
})
export class AppModule {}
```

You can also use the `useClass` syntax:

```typescript
@Module({
  imports: [
    ThrottlerModule.forRootASync({
      imports: [ConfigModule],
      useClass: ThrottlerConfigService,
    }),
  ],
})
export class AppModule {}
```

This is doable, as long as `ThrottlerConfigService` implements the interface `ThrottlerOptionsFactory`.

#### Storages

The built in storage is an in memory cache that keeps track of the requests made until they have passed the TTL set by the global options. You can drop in your own storage option to the `storage` option of the `ThrottlerModule` so long as the class implements the `ThrottlerStorage` interface. 

> > **Note** `ThrottlerStorage` can be imported from `@nestjs/throttler`.

Reviewer 2: ```suggestion
> info **Note** `ThrottlerStorage` can be imported from `@nestjs/throttler`.
```

### Rate Limiting

> In this chapter we cover how to add rate limiting to your application in a more "NestJS Way" than using a package like `express-rate-limit`. To get started, you'll need to install the `@nestjs/throttler` package.

Reviewer 2: A few notes:
1. I'd remove any references to "express" compatible packages
2. Since it's NestJS documentation (and `@nestjs/{..}` package), we should remove "in a more NestJS Way" as this is kinda obvious 😄 
3. Maybe we should leave the first sentence as it was just to let people know what this chapter is about (`A common technique to protect applications from brute-force attacks is **rate-limiting**.`)

```bash
$ npm i --save @nestjs/throttle
```

> Once installation is complete, the `ThrottlerModule` can be configured as any other Nest package with `forRoot` or `forRootAsync`.

Reviewer 2: ```suggestion
Once the installation is complete, the `ThrottlerModule` can be configured as any other Nest package with `forRoot` or `forRootAsync` methods.
```

```typescript
@Module({
  imports: [
    ThrottlerModule.forRoot({
      ttl: 60,
      limit: 10,
    }),
  ]
})
export class AppModule {}
```

> or

Reviewer 2: Following other chapters, can we add a new "Async configuration" section and move this to the end?

```

#### Customization

There may be a time where you want to bind the guard to a controller or globally, but want to avoid rate limiting one or more of your endpoints. For that, you can use the `@SkipThrottle()` decorator, to negate the throttler for an entire class or a single route. The `@SkipThrottle()` decorator can also take in a boolean for if there is a case where you want to exclude _most_ of a controller, but not every route.

There is also the `@Throttle()` decorator which can be used to override the `limit` and `ttl` set in the global module, to give tighter or looser security options. This decorator can be used on a class or a function as well. The order for this decorator does matter, as the arguments are in the order of `limit` `ttl`.

#### Websockets

This module does work with websockets as well, with some limited functionality. First of all, user agent headers are not taken into consideration due to the difference in the underlying transport layer of Socket.IO vs Websockets. The other thing to make note of is that globally bound guards do not activate on websocket gateways, so you **must** bind the guard to the gateway itself using `@UseGuards()`.

#### GraphQL

> Currently, only GraphQL with Express is supported, but Fastify support is coming as well. This module makes use of setting headers through the `res` object and reading headers through the `req` object of Express. To make sure these are available, when configuring your GraphQLModule, make sure the option `context: ({{ '{' }} req, res {{ '}' }}) => ({{ '{' }} req, res {{ '}' }})` is set.

Reviewer 2: Btw `req` should be automatically set by the `@nestjs/graphql` package.

Author: `req` is there by default, but `res` needs to be added for this module, so I thought we should be explicit in adding it. [Though it's looking like Apollo adds `res` by default too](https://www.apollographql.com/docs/apollo-server/api/apollo-server/#middleware-specific-context-fields). Should we keep this, or make a note that this _should_ be default behavior?

#### Configuration

The following options are valid for the `ThrottlerModule`

* ttl: the number of seconds that each request will last in storage
* limit: the maximum number of requests within the TTL limit
* ignoreUserAgents: an array of regular expressions of user-agents to ignore when it comes to throttling requests

> * storage: the storage setting for how to keep track of the requests.

Reviewer 2: Following other chapters, maybe we could wrap it in a table? Example https://docs.nestjs.com/microservices/redis#options
